flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -4049,6 +4097,37 @@ class ArrayExprLowering {
     }
   }
 
+  /// Lower an elemental function array argument with a given
+  /// ConstituentSemantics.
+  template <typename A>
+  CC genElementalArgumentAs(const A &x, ConstituentSemantics semantics) {
+    // Ensure the returned element is in memory if this is what was requested.
+    if ((semantics == ConstituentSemantics::RefOpaque ||
+         semantics == ConstituentSemantics::DataAddr ||
+         semantics == ConstituentSemantics::ByValueArg)) {
+      if (!Fortran::evaluate::IsVariable(x)) {
+        PushSemantics(ConstituentSemantics::DataValue);
+        CC cc = genarr(x);
+        mlir::Location loc = getLoc();
+        if (isParenthesizedVariable(x)) {
+          // Parenthesised variables are lowered to a reference to the variable
+          // storage. When passing it as an argument, a copy must be passed.
+          return [=](IterSpace iters) -> ExtValue {
+            return createInMemoryScalarCopy(builder, loc, cc(iters));
+          };
+        }
+        llvm::Optional<mlir::Type> forceTypeIfLogical =
+            genTypeIfLogical(converter, x);
+        return [=](IterSpace iters) -> ExtValue {
+          return placeScalarValueInMemory(builder, loc, cc(iters),
+                                          forceTypeIfLogical);
```

Comment:
I'm not sure why there is special code for LOGICALs. A boolean value is either true or false and the elemental function has a signature. So it would seem every argument has to be converted to a correctly typed argument. LOGICAL shouldn't be special and require extra heroics any more than computing a `i8` value and passing it as a `ref<i64>`. That is, converting an `i1` to a `!fir.logical<4>`, storing it to a temp, and passing the `ref<!fir.logical<4>>` must be sufficient or else something fundamental is broken.

Reply:
Fair point, I reproduced the logic that was in `genref`, but it is a bit too specific. I have update the logic to just pass the type in all cases.